packages/mobx-state-tree/src/types/utility-types/optional.ts

```diff
@@ -181,20 +170,66 @@ export function optional<IT extends IAnyType>(
             typecheckInternal(type, defaultValueOrFunction)
         }
     }
+}
 
-    const ret = new OptionalValue(type, defaultValueOrFunction)
-    return ret as any
+export function optional<IT extends IAnyType>(
+    type: IT,
+    defaultValueOrFunction: OptionalDefaultValueOrFunction<IT>
+): IOptionalIType<IT, [undefined]>
+export function optional<IT extends IAnyType, OptionalVals extends ValidOptionalValues>(
+    type: IT,
+    defaultValueOrFunction: OptionalDefaultValueOrFunction<IT>,
+    optionalValues: OptionalVals
+): IOptionalIType<IT, OptionalVals>
+/**
+ * `types.optional` - Can be used to create a property with a default value.
+ * If the given value is not provided in the snapshot (or matches one of the provided optional values), it will default to the provided `defaultValue`.
+ * If `defaultValue` is a function, the function will be invoked for every new instance.
+ * Applying a snapshot in which the optional value is _not_ present causes the value to be reset
```

Comment:
I think this API is way better! Super neat!

To see if I understand description correctly:

```javascript
const t = types.model({
  x: types.optional(types.number, 3, [0, 1])
})

t.create({x : 0 }) // x = 3 (would expect 0 though, since it matches types.number?)
t.create({ x: undefined }) // error (not missing, not in the optional set) (should `undefined` be implied always?)
t.create({ }) // x = 3
```

Reply:
optional is basically a snapshot pre-processor :thinking:

Reply:
except that if when used as an snapshot pre-processor (such as the old implementation) then this:
```
const m = M.create({
  x: undefined
})
```
was still "valid" and transformed into the default value, since in the inside it transformed 0 || 1 into undefined for types.optional, and types.optional would accept that undefined

Reply:
A more clear example
```
const M = types.model({ 
  x: types.optionalNull(types.number, 10),
  y: types.optionalNull(types.number, 10)
})

// with the old implementation that uses only a pre-snapshot processor + types.optional
const m = M.create({
  x : undefined  // NOT ok, x will become 10, since it only transforms null into undefined, and undefined will become 10
  y: null // ok, will become undefined then undefined will become 10
})
// with the new implementation
const m = M.create({
  x : undefined, // ok, will throw that undefined is not assignable to number
  y: null // ok, y will become 10
})
```

Reply:
```
const t = types.model({
  x: types.optional(types.number, 3, [0, 1])
})

t.create({x : 0 })
// x = 3, since 3 is an optional value (imagine that for some reason -1 has to be turned into an empty string or whatever, who knows

t.create({ x: undefined })
// error (and TS error) (not missing, not in the optional set), undefined shouldn't be implied since it is not declared as a valid optional value, if wanted then the array could just be [0, 1, undefined]
// this allows to validate stuff such as that the back-end value can be null, but if undefined that's an error

t.create({ }) // error, not 0 or 1 or number, if this should work then use [ 0, 1, undefined ] instead (this will also be a TS error if undefined is not a valid value)
```

Reply:
So answering the question, optional is a type pre-processor but with input snapshot validation on top I guess :)

Reply:
Added clarifications to the function docs